File: server/src/weaverbird/pipeline/steps/pivot.py
from typing import Literal

from weaverbird.pipeline.steps.utils.base import BaseStep
from weaverbird.pipeline.steps.utils.render_variables import StepWithVariablesMixin
from weaverbird.pipeline.types import TemplatedVariable


class PivotStep(BaseStep):
    name: Literal["pivot"] = "pivot"
    index: list[str]
    column_to_pivot: str
    value_column: str
    agg_function: Literal["sum", "avg", "count", "min", "max"]


class PivotStepWithVariable(PivotStep, StepWithVariablesMixin):
    index: TemplatedVariable | list[TemplatedVariable]

File: ui/src/components/stepforms/schemas/pivot.ts
export default {
  $schema: 'http://json-schema.org/draft-07/schema#',
  title: 'Pivot step',
  type: 'object',
  properties: {
    name: {
      type: 'string',
      enum: ['pivot'],
    },
    index: {
      type: 'array',
      items: {
        type: 'string',
        minLength: 1,
      },
      title: 'Index',
      description: 'The columns name used as index',
      attrs: {
        placeholder: 'Select columns',
      },
    },
    columnToPivot: {
      type: 'string',
      minLength: 1,
      title: 'Column to pivot',
      description: 'The column to pivot',
      attrs: {
        placeholder: 'Select a column',
      },
    },
    valueColumn: {
      type: 'string',
      minLength: 1,
      title: 'Value colum...',
      description: 'Column on which aggregate function will apply',
      attrs: {
        placeholder: 'Select a column',
      },
    },
    aggFunction: {
      type: 'string',
      enum: ['sum', 'avg', 'count', 'min', 'max'],
    },
  },
  required: ['name', 'index', 'columnToPivot', 'valueColumn', 'aggFunction'],
  additionalProperties: false,
};
